cpp/include/cudf/algorithm/scan_copy_if.cuh

```diff
@@ -0,0 +1,440 @@
+
+    // Scatter
+
+    for (uint32_t i = 0; i < Policy::ITEMS_PER_THREAD; ++i) {
+      if (selection_flags[i]) {
+        if (selection_indices[i] < num_selections) {  //
+          d_output[selection_indices[i]] =
+            thrust::make_pair(tile_offset + threadIdx.x * Policy::ITEMS_PER_THREAD + i, items[i]);
+        }
+      }
+    }
+
+    return num_selections;
+  }
+};
+
+// ===== Kernels ===================================================================================
+
+template <typename Policy>
+__global__ void initialization_pass_kernel(  //
+  typename Policy::ItemsTileState items_state,
+  typename Policy::IndexTileState index_state,
+  uint32_t num_tiles)
+{
+  items_state.InitializeStatus(num_tiles);
+  index_state.InitializeStatus(num_tiles);
+}
+
+template <typename Policy>
+__global__ void execution_pass_kernel(  //
+  typename Policy::InputIterator d_input,
+  typename Policy::OutputCountIterator d_output_count,
+  typename Policy::OutputIterator d_output,
+  uint32_t num_items,
+  typename Policy::ScanOperator scan_op,
+  typename Policy::PredOperator pred_op,
+  typename Policy::ItemsTileState items_tile_state,
+  typename Policy::IndexTileState index_tile_state,
+  bool do_scatter,
+  uint32_t num_tiles,
+  uint32_t start_tile)
+{
+  auto agent_instance = agent<Policy>{
+    d_input,
+    d_output_count,
+    d_output,
+    num_items,
+    scan_op,
+    pred_op,
+    items_tile_state,
+    index_tile_state  //
+  };
+
+  agent_instance.consume_range(do_scatter, num_tiles, start_tile);
+}
+
+// ===== Policy ====================================================================================
+
+template <typename InputIterator_,
+          typename OutputCountIterator_,
+          typename OutputIterator_,
+          typename ScanOperator_,
+          typename PredicateOperator_>
+struct policy {
+  static constexpr uint32_t THREADS_PER_INIT_BLOCK = 128;
+  static constexpr uint32_t THREADS_PER_BLOCK      = 128;
+  static constexpr uint32_t ITEMS_PER_THREAD       = 16;
+  static constexpr uint32_t ITEMS_PER_TILE         = ITEMS_PER_THREAD * THREADS_PER_BLOCK;
+
+  using InputIterator       = InputIterator_;
+  using OutputIterator      = OutputIterator_;
+  using OutputCountIterator = OutputCountIterator_;
+
+  using ScanOperator = ScanOperator_;
+  using PredOperator = PredicateOperator_;
+
+  using Offset      = typename std::iterator_traits<InputIterator>::difference_type;
+  using Input       = typename std::iterator_traits<InputIterator>::value_type;
+  using OutputCount = typename std::iterator_traits<OutputCountIterator>::value_type;
+  using OutputValue = typename std::iterator_traits<OutputIterator>::value_type;
+
+  // Items Load
+
+  using ItemsBlockLoad = cub::BlockLoad<  //
+    Input,
+    THREADS_PER_BLOCK,
+    ITEMS_PER_THREAD,
+    cub::BlockLoadAlgorithm::BLOCK_LOAD_DIRECT>;
+
+  // Items Scan
+
+  using ItemsTileState      = cub::ScanTileState<Input>;
```

Comment:
You might want to use copies of these instead of reusing CUB's helper classes. I reserve the right to change the `agent` layer of CUB at anytime, since those are the performance-critical backends for the device algorithms.

Alternatively, is this algorithm general purpose enough that it should be added to CUB directly? Can you give me a high-level description of what it does?